File: infrastructures/ddd/csharp/src/Infrastructure/Queues/Infra/IEventPublisher.cs
using System.Threading.Tasks;

namespace OddlyDdd.Infrastructure.Queues.Infra
{
    /// <summary>
    /// Interface for publishing domain events to a message queue.
    /// Used for asynchronous communication between subdomains.
    /// CRITICAL: Domain events via queue are the ONLY way to communicate between subdomains.
    /// HTTP calls between subdomains are FORBIDDEN.
    /// </summary>
    public interface IEventPublisher
    {
        /// <summary>
        /// Publishes a domain event to the specified topic.
        /// Topic naming convention: {subdomain}.{action}
        /// Examples: "example.created", "example.updated", "example.deleted"
        /// </summary>
        /// <typeparam name="TEvent">The event type</typeparam>
        /// <param name="p_event">The event to publish</param>
        /// <param name="p_topic">The topic to publish to</param>
        Task PublishAsync<TEvent>(TEvent p_event, string p_topic);
    }
}

File: infrastructures/ddd/typescript/src/infrastructure/queues/infra/IEventPublisher.ts
/**
 * Interface for publishing domain events to a message queue.
 * Event publishers:
 * - Publish domain events to topics
 * - Used for subdomain-to-subdomain communication
 * - ONLY way to communicate between subdomains (NO HTTP calls)
 * 
 * Topic naming convention: {subdomain}.{action}
 * Examples: "example.created", "example.updated", "example.deleted"
 */
export interface IEventPublisher {
  /**
   * Publishes a domain event to the specified topic.
   * @param p_event The event to publish
   * @param p_topic The topic to publish to
   */
  PublishAsync<TEvent>(p_event: TEvent, p_topic: string): Promise<void>;
}
